### PR TITLE
4.0 dev

### PR DESCRIPTION
**Pull request by Alex-Werner**
## Issue being fixed or feature implemented

This version fixes any previous issue browser user might have had on browser. 
It cames with some breaking changes : 
- Usage of wallet-lib has changed
- Initialization of Dash is fastest by defer all pre-emptive async work to when they are needed.

In addition, many improvements in process has been added (karma, typings,...)

## What was done?

- fix: prefetch contract on Client's constructor initialization (#64) 
- tests: added karma and functional browser test (#63) 
- doc: client params (#62)
- impr: error handling & rejections (#59)

## How Has This Been Tested?

This PR brings additional tests.

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: src/SDK/Client/Platform/broadcastStateTransition.ts [hunk 1]
```diff
@@ -23,6 +25,7 @@ export default async function broadcastStateTransition(platform: Platform, state
     // if (!result.isValid()) {
     //     throw new Error(`StateTransition is invalid - ${JSON.stringify(result.getErrors())}`);
     // }
+    const { client: dapiClient } = await client.getDAPIClient();
```
Comment (shumkov) [inline src/SDK/Client/Platform/broadcastStateTransition.ts:28]: This looks strange... It supposed to be sync and respond with DAPI Client instance.

File: src/SDK/Client/Platform/methods/contracts/get.spec.ts [hunk 2]
```diff
@@ -14,17 +14,20 @@ const factory = new DataContractFactory(
...
-            // @ts-ignore
-            case contractsFixtures.ratePlatform.id:
-                const contract = await dpp.dataContract.createFromObject(contractsFixtures.ratePlatform);
-                return contract.serialize()
-            default:
-                return null;
-        }
-    },
+    getDAPIClient: () => {
+        return { client: { getDataContract } };
```
Comment (shumkov) [inline src/SDK/Client/Platform/methods/contracts/get.spec.ts:29]: Just `{  getDataContract  };` ?

File: src/SDK/Client/Platform/broadcastStateTransition.ts (hunk 1 above)
Comment (Alex-Werner) [reply src/SDK/Client/Platform/broadcastStateTransition.ts:28]: The bigger issue is : DAPIClient vs DAPIClientWrapper. 

In DAPIClient, elements are not the object we look for thus the need for the wrapper. 
I agree it's an issue, I propose renaming walletLib.transporters.DAPIClient to DAPIClientWrapper.

File: src/SDK/Client/Platform/methods/names/get.ts [hunk 3]
```diff
@@ -16,7 +16,7 @@ export async function get(this: Platform, id: string): Promise<any> {
     try{
         const documents = await this.documents.get('dpns.domain', queryOpts);
         return (documents[0] !== undefined) ? documents[0] : null;
-    }catch (e) {
+    } catch (e) {
```
Comment (shumkov) [inline src/SDK/Client/Platform/methods/names/get.ts:19]: Does this `try/catch` make sense?

File: src/SDK/Client/Platform/methods/contracts/get.spec.ts (hunk 2 above)
Comment (Alex-Werner) [reply src/SDK/Client/Platform/methods/contracts/get.spec.ts:29]: Yeah, no need to mock more on this one, only getDataContract is needed to get a contract =D

File: src/SDK/Client/Platform/methods/names/get.ts (hunk 3 above)
Comment (Alex-Werner) [reply src/SDK/Client/Platform/methods/names/get.ts:19]: Absolutely not ! 
Removing that non-sense right away.

File: src/SDK/Client/Platform/methods/identities/register.ts [hunk 4]
```diff
@@ -31,76 +29,72 @@ export async function register(this: Platform): Promise<any> {
...
-        // @ts-ignore
-        return identity;
-    } catch (e) {
-        console.error(`Identity registration failed:`,e);
-        throw e;
+    if (!result.isValid()) {
+        throw new Error(`StateTransition is invalid - ${JSON.stringify(result.getErrors())}`);
     }
+
+    const { client: dapiClient } = await this.client.getDAPIClient();
```
Comment (shumkov) [inline src/SDK/Client/Platform/methods/identities/register.ts:92]: Not sure that it's correct

File: src/SDK/Client/Platform/methods/identities/get.ts [hunk 5]
```diff
@@ -2,17 +2,20 @@ import {Platform} from "../../Platform";
...
+ *
  * @param {Platform} this - bound instance class
  * @param {string} id - id
  * @returns identites
  */
 export async function get(this: Platform, id: string): Promise<any> {
-    // @ts-ignore
-    const identityBuffer = await this.client.getIdentity(id);
-    if(identityBuffer===null){
+    const { client: dapiClient } = await this.client.getDAPIClient();
```
Comment (shumkov) [inline src/SDK/Client/Platform/methods/identities/get.ts:11]: Not sure that's correct.

File: src/SDK/Client/Platform/methods/documents/get.ts [hunk 6]
```diff
@@ -36,24 +53,22 @@ export async function get(this: Platform, typeLocator: string, opts: fetchOpts):
     if (!app.contractId) {
         throw new Error(`Missing contract ID for ${appName}`)
     }
+
     const contractId = app.contractId;
-    try{
+    try {
+        // If not present, will fetch contract based on appName and contractId store in this.apps.
+        await ensureAppContractFetched.call(this, appName);
+        const { client: dapiClient } = await this.client.getDAPIClient();
```
Comment (shumkov) [inline src/SDK/Client/Platform/methods/documents/get.ts:61]: Not sure that's correct.

File: src/SDK/Client/Platform/methods/documents/get.ts [hunk 7]
```diff
@@ -36,24 +53,22 @@ export async function get(this: Platform, typeLocator: string, opts: fetchOpts):
...
-                console.error('Document creation: failure', e);
-                throw e;
-            }
+            const doc = await this.dpp.document.createFromSerialized(rawData, {skipValidation: true});
+            documents.push(doc);
         }
         return documents
     } catch (e) {
-        console.error(`Document creation: unable to get documents of ${contractId}`);
         throw e;
```
Comment (shumkov) [inline src/SDK/Client/Platform/methods/documents/get.ts:72]: Let's remove this `try/catch`

File: src/SDK/Client/Platform/methods/contracts/get.ts [hunk 8]
```diff
@@ -11,6 +11,7 @@ declare type ContractIdentifier = string;
  */
 export async function get(this: Platform, identifier: ContractIdentifier): Promise<any> {
     let localContract;
+    const { client: dapiClient } = await this.client.getDAPIClient();
```
Comment (shumkov) [inline src/SDK/Client/Platform/methods/contracts/get.ts:14]: Not sure that's correct.

File: src/SDK/Client/Platform/methods/names/get.ts (hunk 3 above)
Comment (Alex-Werner) [reply src/SDK/Client/Platform/methods/names/get.ts]: Found 3 instances that I removed

File: src/SDK/Client/Platform/methods/documents/get.ts (hunk 7 above)
Comment (Alex-Werner) [reply src/SDK/Client/Platform/methods/documents/get.ts]: Done

**Pull request merged**